clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThread.java

```diff
@@ -325,4 +330,23 @@ void cleanup() {
             log.debug("Closed the consumer network thread");
         }
     }
+
+    /**
+     * If there is a metadata error, complete the completable events which are not passed by 
+     * error event with the metadata error
+     * @param events the completable events
+     */
+    private void maybeFailOnMetadataError(List<CompletableEvent<?>> events) {
+        if (networkClientDelegate.metadataError().isPresent()) {
+            Throwable metadataError = networkClientDelegate.metadataError().get();
+            List<CompletableEvent<?>> completableApplicationEvent = events.stream()
+                    .filter(event -> !(event instanceof CompletableApplicationEvent && 
+                            ((CompletableApplicationEvent<?>) event).isPassedByErrorEvent()))
+                    .collect(Collectors.toList());
+            if (!completableApplicationEvent.isEmpty()) {
+                completableApplicationEvent.forEach(event -> event.future().completeExceptionally(metadataError));
+                networkClientDelegate.clearMetadataError();
+            }
+        }
+    }
```

Comment:
It seems like this could be a little more succinct:

```suggestion
    private void maybeFailOnMetadataError() {
       networkClientDelegate.metadataError().ifPresent(metadataError -> {
            applicationEventReaper.tracked().stream()
                .filter(e -> e instanceof CompletableApplicationEvent<?>)
                .map(e -> (CompletableApplicationEvent<?>) e)
                .filter(e -> e.isPassedByErrorEvent())
                .filter(e -> !e.future().isDone())
                .forEach(e -> e.future().completeExceptionally(metadataError));
            networkClientDelegate.clearMetadataError();
        }));
    }
```

Reply:
Do we only want to clear the error if an event "consumed" the error?

Reply:
If no events are present, we should not clear the metadata error. Instead, we should propagate this error to the next event. Therefore, using `foreach` to handle this scenario is not appropriate. we should check it have any event first.